File: Data Structures & Algorithms/binary-tree-right-side-view/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        temp=[]
        depth=0
        res=[]
        def dfs(root,depth):
            if not root:
                return None
            if len(temp)==depth:
                temp.append([])
            temp[depth].append(root.val)
            dfs(root.left,depth+1)
            dfs(root.right,depth+1)
        dfs(root,0)
        for i in temp:
            res.append(i[-1])
        return res

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-5.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        String t=serialise(root);
        String p=serialise(subRoot);
        int i=1,j=0;
        int lps[]=new int[p.length()];
        while(i<p.length()){
            if(p.charAt(j) == p.charAt(i)){
                lps[i]=j+1;
                i++;
                j++;
            }
            else if(j == 0) i++;
            else j=lps[j-1];
        }
        i=0;j=0;
        while(i<t.length()){
            if(t.charAt(i) == p.charAt(j)){
                i++;
                j++;
            }
            else{
                if(j == 0) i++;
                else j=lps[j-1];
            }
            if(j == p.length()) return true;
        }
        return false;
    }
    public String serialise(TreeNode node){
        if(node == null) return "# ";
        return node.val + " " + serialise(node.left) + serialise(node.right);
    }
}
